In [0]:
# Use Google Colab GPU in Task3
import pandas as pd
import numpy as np
from scipy.io import loadmat
import keras

### Read and Preprocess data

In [0]:
traindata = loadmat('train_32x32.mat')
testdata = loadmat('test_32x32.mat')
x_train = traindata['X']
y_train = traindata['y']
x_test = testdata['X']
y_test = testdata['y']

In [0]:
# y_train,y_test changed to (0,9)
y_train = y_train -1
y_test = y_test -1

# normalize X
x_train = x_train/255
x_test = x_test/255

In [0]:
num_classes = 10
#input image dimensions
img_rows, img_cols = 32, 32

x_train_images = x_train.transpose(3, 0, 1, 2)
x_test_images = x_test.transpose(3, 0, 1, 2)
input_shape = (img_rows, img_cols, 3)

y_train = keras.utils.to_categorical(y_train, num_classes)
y_test = keras.utils.to_categorical(y_test, num_classes)

### Train convolutional neural network with a base model

In [0]:
from keras.layers import Conv2D, MaxPooling2D, Flatten, Dense, Activation
from keras.models import Sequential
cnn = Sequential()
cnn.add(Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=input_shape))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Conv2D(32, (3, 3), activation='relu'))
cnn.add(MaxPooling2D(pool_size=(2, 2)))
cnn.add(Flatten())
cnn.add(Dense(64, activation='relu'))
cnn.add(Dense(num_classes, activation='softmax'))

In [36]:
cnn.compile("adam", "categorical_crossentropy", metrics=['accuracy'])
history_cnn = cnn.fit(x_train_images, y_train,
                      batch_size=128, epochs=10, verbose=1, validation_split=.1)
cnn.evaluate(x_test_images, y_test)

Train on 65931 samples, validate on 7326 samples
Epoch 1/10
65931/65931 [==============================] - 11s 172us/step - loss: 1.2217 - acc: 0.6102 - val_loss: 0.6680 - val_acc: 0.8070
Epoch 2/10
65931/65931 [==============================] - 9s 138us/step - loss: 0.6172 - acc: 0.8260 - val_loss: 0.6174 - val_acc: 0.8281
Epoch 3/10
35712/65931 [===============>..............] - ETA: 3s - loss: 0.5535 - acc: 0.846165931/65931 [==============================] - 9s 138us/step - loss: 0.5436 - acc: 0.8489 - val_loss: 0.5252 - val_acc: 0.8546
Epoch 4/10
65931/65931 [==============================] - 9s 137us/step - loss: 0.4988 - acc: 0.8619 - val_loss: 0.4944 - val_acc: 0.8628
Epoch 5/10
65931/65931 [==============================] - 9s 138us/step - loss: 0.4663 - acc: 0.8707 - val_loss: 0.4742 - val_acc: 0.8720
Epoch 6/10
 5120/65931 [=>............................] - ETA: 7s - loss: 0.4609 - acc: 0.869165931/65931 [==============================] - 9s 138us/step - loss: 0.4410 - acc: 

[0.503243366596372, 0.8638982790411801]

#### As reported above, the base model contained two convolutional layers with epochs = 10, and got the final validation accuracy of 0.8791 and test accuracy of 0.8639. higher than 0.85.

### Using Batch Normalization

In [0]:
from keras.layers import BatchNormalization
cnn_small_bn = Sequential()
cnn_small_bn.add(Conv2D(32, kernel_size=(3, 3),
                 input_shape=input_shape))
cnn_small_bn.add(Activation("relu"))
cnn_small_bn.add(BatchNormalization())
cnn_small_bn.add(MaxPooling2D(pool_size=(2, 2)))
cnn_small_bn.add(Conv2D(32, (3, 3)))
cnn_small_bn.add(Activation("relu"))
cnn_small_bn.add(BatchNormalization())
cnn_small_bn.add(MaxPooling2D(pool_size=(2, 2)))
cnn_small_bn.add(Flatten())
cnn_small_bn.add(Dense(64, activation='relu'))
cnn_small_bn.add(Dense(num_classes, activation='softmax'))

In [44]:
cnn_small_bn.compile("adam", "categorical_crossentropy", metrics=['accuracy'])
history_cnn_small_bn = cnn_small_bn.fit(x_train_images, y_train,
                      batch_size=128, epochs=50, verbose=1, validation_split=.1)
cnn_small_bn.evaluate(x_test_images, y_test)

Train on 65931 samples, validate on 7326 samples
Epoch 1/50
65931/65931 [==============================] - 13s 197us/step - loss: 0.8242 - acc: 0.7444 - val_loss: 0.5338 - val_acc: 0.8425
Epoch 2/50
65931/65931 [==============================] - 12s 183us/step - loss: 0.4217 - acc: 0.8754 - val_loss: 0.4457 - val_acc: 0.8698
Epoch 3/50
15744/65931 [======>.......................] - ETA: 8s - loss: 0.3483 - acc: 0.899865931/65931 [==============================] - 12s 182us/step - loss: 0.3498 - acc: 0.8976 - val_loss: 0.4500 - val_acc: 0.8694
Epoch 4/50
65931/65931 [==============================] - 12s 184us/step - loss: 0.3038 - acc: 0.9100 - val_loss: 0.4583 - val_acc: 0.8610
Epoch 5/50
46208/65931 [====================>.........] - ETA: 3s - loss: 0.2709 - acc: 0.919465931/65931 [==============================] - 12s 183us/step - loss: 0.2720 - acc: 0.9185 - val_loss: 0.4157 - val_acc: 0.8802
Epoch 6/50
65931/65931 [==============================] - 12s 185us/step - loss: 0.2417 - 

53376/65931 [=======================>......] - ETA: 2s - loss: 0.0571 - acc: 0.984865931/65931 [==============================] - 12s 181us/step - loss: 0.0536 - acc: 0.9855 - val_loss: 0.8538 - val_acc: 0.8817
Epoch 48/50
65931/65931 [==============================] - 12s 182us/step - loss: 0.0201 - acc: 0.9953 - val_loss: 0.8431 - val_acc: 0.8893
Epoch 49/50
54656/65931 [=======================>......] - ETA: 1s - loss: 0.0157 - acc: 0.997065931/65931 [==============================] - 12s 183us/step - loss: 0.0160 - acc: 0.9967 - val_loss: 0.8591 - val_acc: 0.8897
Epoch 50/50
26032/26032 [==============================] - 5s 177us/step


[1.0841373216244035, 0.8656653349723418]

#### Then we added the Batch Normalization to the base model, and the test accuracy increased to 0.8657.

## Complicated Model with Dropout and more epochs

#### In the following part, we tried two models to improve our test accuracy and used more epochs (e.g. 50).   
#### Model 1 used three convolutional layers with Dropout which reached the test accuracy of 0.9061. In model 2, each layer contained two convolutional sublayers and one maxpooling sublayer. It reached the test accuracy of 0.8911.  
#### In conclusion, our final test accuracy increased to 0.9061.

### Model 1

In [0]:
# three layers with Dropout
from keras.layers import BatchNormalization, Dropout
cnn_bn = Sequential()
cnn_bn.add(Conv2D(32, kernel_size=(3, 3),
                 input_shape=input_shape))
cnn_bn.add(Activation("relu"))
cnn_bn.add(BatchNormalization())
cnn_bn.add(MaxPooling2D(pool_size=(2, 2)))
cnn_bn.add(Conv2D(32, (3, 3)))
cnn_bn.add(Activation("relu"))
cnn_bn.add(BatchNormalization())
cnn_bn.add(MaxPooling2D(pool_size=(2, 2)))
cnn_bn.add(Conv2D(32, (3, 3)))
cnn_bn.add(Activation("relu"))
cnn_bn.add(BatchNormalization())
cnn_bn.add(MaxPooling2D(pool_size=(2, 2)))
cnn_bn.add(Flatten())
cnn_bn.add(Dense(64, activation='relu'))
cnn_bn.add(Dropout(0.5))
cnn_bn.add(Dense(num_classes, activation='softmax'))

In [40]:
cnn_bn.compile("adam", "categorical_crossentropy", metrics=['accuracy'])
history_cnn_bn = cnn_bn.fit(x_train_images, y_train,
                      batch_size=128, epochs=50, verbose=1, validation_split=.1)
cnn_bn.evaluate(x_test_images, y_test)

Train on 65931 samples, validate on 7326 samples
Epoch 1/50
65931/65931 [==============================] - 14s 218us/step - loss: 1.3335 - acc: 0.5621 - val_loss: 0.6467 - val_acc: 0.8008
Epoch 2/50
65931/65931 [==============================] - 14s 205us/step - loss: 0.6892 - acc: 0.7893 - val_loss: 0.5428 - val_acc: 0.8310
Epoch 3/50
 5760/65931 [=>............................] - ETA: 11s - loss: 0.6220 - acc: 0.809265931/65931 [==============================] - 14s 205us/step - loss: 0.5660 - acc: 0.8275 - val_loss: 0.5236 - val_acc: 0.8402
Epoch 4/50
65931/65931 [==============================] - 13s 203us/step - loss: 0.5046 - acc: 0.8499 - val_loss: 0.4194 - val_acc: 0.8735
Epoch 5/50
31616/65931 [=============>................] - ETA: 6s - loss: 0.4744 - acc: 0.860665931/65931 [==============================] - 13s 204us/step - loss: 0.4661 - acc: 0.8623 - val_loss: 0.4018 - val_acc: 0.8797
Epoch 6/50
65931/65931 [==============================] - 14s 207us/step - loss: 0.4371 -

39168/65931 [================>.............] - ETA: 5s - loss: 0.2353 - acc: 0.929065931/65931 [==============================] - 14s 205us/step - loss: 0.2415 - acc: 0.9265 - val_loss: 0.3313 - val_acc: 0.9085
Epoch 48/50
65931/65931 [==============================] - 13s 203us/step - loss: 0.2399 - acc: 0.9277 - val_loss: 0.3344 - val_acc: 0.9057
Epoch 49/50
41856/65931 [==================>...........] - ETA: 4s - loss: 0.2414 - acc: 0.927365931/65931 [==============================] - 13s 202us/step - loss: 0.2414 - acc: 0.9277 - val_loss: 0.3620 - val_acc: 0.9027
Epoch 50/50
26032/26032 [==============================] - 5s 187us/step


[0.36984329247422465, 0.9061539643515673]

### Model 2

In [0]:
# conv,relu,conv,relu,pool, two layers
from keras.layers import BatchNormalization
cnn_bn_c = Sequential()
cnn_bn_c.add(Conv2D(32, kernel_size=(3, 3),
                 input_shape=input_shape))
cnn_bn_c.add(Activation("relu"))
cnn_bn_c.add(Conv2D(32, kernel_size=(3, 3)))
cnn_bn_c.add(Activation("relu"))
cnn_bn_c.add(BatchNormalization())
cnn_bn_c.add(MaxPooling2D(pool_size=(2, 2)))
cnn_bn_c.add(Conv2D(32, (3, 3)))
cnn_bn_c.add(Activation("relu"))
cnn_bn_c.add(Conv2D(32, (3, 3)))
cnn_bn_c.add(Activation("relu"))
cnn_bn_c.add(BatchNormalization())
cnn_bn_c.add(MaxPooling2D(pool_size=(2, 2)))
cnn_bn_c.add(Flatten())
cnn_bn_c.add(Dense(64, activation='relu'))
cnn_bn_c.add(Dense(num_classes, activation='softmax'))

In [42]:
cnn_bn_c.compile("adam", "categorical_crossentropy", metrics=['accuracy'])
history_cnn_bn_c = cnn_bn_c.fit(x_train_images, y_train,
                      batch_size=128, epochs=50, verbose=1, validation_split=.1)
cnn_bn_c.evaluate(x_test_images, y_test)

Train on 65931 samples, validate on 7326 samples
Epoch 1/50
65931/65931 [==============================] - 17s 256us/step - loss: 0.7357 - acc: 0.7693 - val_loss: 0.6565 - val_acc: 0.8033
Epoch 2/50
56704/65931 [========================>.....] - ETA: 2s - loss: 0.3826 - acc: 0.885965931/65931 [==============================] - 15s 233us/step - loss: 0.3784 - acc: 0.8877 - val_loss: 0.4071 - val_acc: 0.8776
Epoch 3/50
65931/65931 [==============================] - 15s 233us/step - loss: 0.3077 - acc: 0.9080 - val_loss: 0.4123 - val_acc: 0.8746
Epoch 4/50
27520/65931 [===========>..................] - ETA: 8s - loss: 0.2544 - acc: 0.922565931/65931 [==============================] - 15s 230us/step - loss: 0.2669 - acc: 0.9203 - val_loss: 0.3726 - val_acc: 0.8848
Epoch 5/50
65931/65931 [==============================] - 15s 232us/step - loss: 0.2389 - acc: 0.9298 - val_loss: 0.3462 - val_acc: 0.8986
Epoch 6/50
20864/65931 [========>.....................] - ETA: 9s - loss: 0.1951 - acc: 0.

18560/65931 [=======>......................] - ETA: 10s - loss: 0.0177 - acc: 0.995065931/65931 [==============================] - 15s 225us/step - loss: 0.0223 - acc: 0.9937 - val_loss: 0.7550 - val_acc: 0.8930
Epoch 47/50
65931/65931 [==============================] - 15s 229us/step - loss: 0.0346 - acc: 0.9897 - val_loss: 0.8504 - val_acc: 0.8867
Epoch 48/50
18304/65931 [=======>......................] - ETA: 10s - loss: 0.0191 - acc: 0.994765931/65931 [==============================] - 15s 226us/step - loss: 0.0225 - acc: 0.9938 - val_loss: 0.8405 - val_acc: 0.8862
Epoch 49/50
65931/65931 [==============================] - 15s 225us/step - loss: 0.0203 - acc: 0.9943 - val_loss: 0.8103 - val_acc: 0.8930
Epoch 50/50
26032/26032 [==============================] - 5s 188us/step


[0.8574302408954443, 0.8910955746773203]